module/zfs/vdev_draid.c

```diff
@@ -0,0 +1,1661 @@
+ *    +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ *    | 6 | 12| 8 | 14| 7 | 10| 13| 2 | 5 | 3 | 1 | 11| 16| 9 | 4 | 15| 0 |
+ * s  |         group 15              |          group 16         |       |
+ * l  |   |           group 17            |        group 18       |       |
+ * i  |       |           group 19            |        group 20   |       |
+ * c  |           |           group 21            |       group 22|       |
+ * e  |               |           group 23            |           |       |
+ * 2  |  group 24         |           group 25            |       |       |
+ *    |      group 26         |           group 27            |   |       |
+ *    |          group 28         |           group 29            |       |
+ *    +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ *
+ *    ...
+ *
+ * Notes:
+ * 1. In this configuration the number of groups equals the number of data
+ *    disks. This will occur but it should never be depended upon.
+ *
+ * 2. The top row of values (5, 11, 7, ...) reflects an example disk
+ *    permutation mapping which has been applied to the slice.
+ *
+ * This layout has several advantages:
+ *
+ * 1. The group count is not a relevant parameter when defining a dRAID
+ *    layout. Only the group size is needed, and *all* groups will have the
+ *    desired size.
+ *
+ * 2. All possible group sizes (<= physical disk count) can be supported.
+ *
+ * 3. The logic within vdev_draid.c is simplified when the group size is
+ *    the same for all groups (although some of the logic around computing
+ *    permutation numbers and drive offsets is more complicated).
+ */
+
+static vdev_t *vdev_draid_spare_get_child(vdev_t *vd, uint64_t offset);
+
+/*
+ * Lookup the permutation array and iteration id for the provided offset.
+ */
+static void
+vdev_draid_get_perm(vdev_draid_config_t *vdc, uint64_t pindex,
+    uint64_t **base, uint64_t *iter)
+{
+	uint64_t ncols = vdc->vdc_children;
+	uint64_t poff = pindex % (vdc->vdc_nperms * ncols);
+
+	*base = vdc->vdc_perms + (poff / ncols) * ncols;
+	*iter = poff % ncols;
+}
+
+static inline uint64_t
+vdev_draid_permute_id(vdev_draid_config_t *vdc,
+    uint64_t *base, uint64_t iter, uint64_t index)
+{
+	return ((base[index] + iter) % vdc->vdc_children);
+}
+
+/*
+ * Full stripe writes.  For "big columns" it's sufficient to map the correct
+ * range of the zio ABD.  Partial columns require allocating a gang ABD in
+ * order to zero fill the skip sectors.  When the column is empty a zero
+ * filled skip sector must be mapped.  In all cases the data ABDs must be
+ * the same size as the parity ABDs.
+ *
+ * Both rm->cols and rc->rc_size are increased to calculate the parity over
+ * the full stripe width.  All zero filled skip sectors must be written to
+ * disk for the benefit of the device rebuild feature which is unaware of
+ * individual block boundaries.
```

Comment:
Can you explain *why* device rebuild requires that the skip sectors be explicitly zero-filled?  Presumably because they are relevant to the parity calculation.

Reply:
Will do.